aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -288,41 +286,41 @@ static void norm_kernel_tensor_iterator_impl(
       });
       return;
     }
-    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
-      using acc_t = typename scalar_value_type<scalar_t>::type;
-      binary_kernel_reduce(
-        iter,
-        NormTwoOps<scalar_t, acc_t>(),
-        acc_t(0)
-      );
-    });
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+        kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
+          using real_type = typename scalar_value_type<scalar_t>::type;
+          using real_acc_t = at::opmath_type<real_type>;
+          binary_kernel_reduce(
+              iter, NormTwoOps<scalar_t, real_acc_t>(), real_acc_t(0));
+        });
   } else if (val == INFINITY) {
-    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
-      using acc_t = typename scalar_value_type<scalar_t>::type;
-      binary_kernel_reduce(
-        iter,
-        AbsMaxOps<scalar_t, acc_t>(),
-        acc_t(0)
-      );
-    });
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+        kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
+          using real_type = typename scalar_value_type<scalar_t>::type;
+          using real_acc_t = at::opmath_type<real_type>;
+          binary_kernel_reduce(
+              iter, AbsMaxOps<scalar_t, real_acc_t>(), real_acc_t(0));
+        });
   } else if (val == -INFINITY) {
-    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
-      using acc_t = typename scalar_value_type<scalar_t>::type;
-      binary_kernel_reduce(
-        iter,
-        AbsMinOps<scalar_t, acc_t>(),
-        std::numeric_limits<acc_t>::infinity()
-      );
-    });
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+        kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
+          using real_type = typename scalar_value_type<scalar_t>::type;
+          using real_acc_t = at::opmath_type<real_type>;
+          binary_kernel_reduce(
+              iter,
+              AbsMinOps<scalar_t, real_acc_t>(),
```

Comment:
Even though it is not really necessary for min/max, it should be no harm to do a cast explicitly anyway since we do comparison after upcasting for bf16 or fp16.